### PR TITLE
feat: [PL-23377]: Adding Azure support to SCM

### PR DESCRIPTION
**Pull request by tphoney**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/31165)

### CODE REVIEW

**Comment by tphoney** [comment]
trigger messagecheck

**Comment by tphoney** [comment]
trigger messagecheck

**Comment by tphoney** [comment]
trigger messagecheck

**Comment by tphoney** [comment]
trigger messagecheck

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31165)* status: Some of the checks are too old. 14 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[product/ci/scm/README.md, line 89 at r1](https://reviewable.io/reviews/harness/harness-core/31165#-MyBEjPtB5kv2pveFeyO:-MyBEjPtB5kv2pveFeyP:b-vdyhbu) ([raw file](https://github.com/harness/harness-core/blob/db40ed36ebdaa0f5a07205875e04f2ee419ff012/product/ci/scm/README.md#L89)):*
> ```Markdown
> 
> ```BASH
> AZURE_TOKEN=963408579168567c07ff8bfd2a5455e5307f74d4 go test -count=1 -v ./...
> ```

Is this a valid token? 
___
*[product/ci/scm/file/file.go, line 452 at r1](https://reviewable.io/reviews/harness/harness-core/31165#-MyBG0Wj2fs0pHmYKBGx:-MyBG0Wj2fs0pHmYKBGy:b-54awv) ([raw file](https://github.com/harness/harness-core/blob/db40ed36ebdaa0f5a07205875e04f2ee419ff012/product/ci/scm/file/file.go#L452)):*
> ```Go
> 			return "", ""
> 		}
> 		return out.RefUpdates[0].NewObjectID, ""
> ```

Shouldn't NewObjectID be 2nd in the return param list? First one is commit id right?

**Comment by tphoney** [comment]
i'll look at these now

File: product/ci/scm/file/file.go [hunk 1]
```diff
@@ -429,12 +433,29 @@ func parseCrudResponse(ctx context.Context, client *scm.Client, body io.Reader,
...
+			} `json:"refUpdates"`
+		}
+		out := azureResponse{}
+		err := json.Unmarshal([]byte(bodyStr), &out)
+		// there is no commit id or sha, no need to error
+		if err != nil {
+			log.Errorw("parseCrudResponse unable to get commitid/blobid from Azure CRUD operation", zap.Error(err))
+			return "", ""
+		}
+		return out.RefUpdates[0].NewObjectID, ""
```
Comment (shubham149) [inline product/ci/scm/file/file.go:452]: Ensure that we don't get null pointer exception here i.e if RefUpdates list is empty

Comment (tphoney) [reply product/ci/scm/file/file.go:454]: done

**Review by tphoney — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31165)* status: 14 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[WORKSPACE, line 1012 at r1](https://reviewable.io/reviews/harness/harness-core/31165#-MxKa5Passo7xBBp557F-r1-1012:-MyDIAY1BNLm0K5h8uqg:bkykn6n) ([raw file](https://github.com/harness/harness-core/blob/db40ed36ebdaa0f5a07205875e04f2ee419ff012/WORKSPACE#L1012)):*
<details><summary><i>Previously, shubham149 (Shubham Agrawal) wrote…</i></summary><blockquote>

Shouldn't this use version v1.20
</blockquote></details>

it was not available at the time, updated now
___
*[product/ci/scm/README.md, line 89 at r1](https://reviewable.io/reviews/harness/harness-core/31165#-MyBEjPtB5kv2pveFeyO:-MyDI2PZCzDXVgXV18V-:b7303el) ([raw file](https://github.com/harness/harness-core/blob/db40ed36ebdaa0f5a07205875e04f2ee419ff012/product/ci/scm/README.md#L89)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

Is this a valid token? 
</blockquote></details>

it is not a valid token, copied from the line above.
___
*[product/ci/scm/file/file.go, line 452 at r1](https://reviewable.io/reviews/harness/harness-core/31165#-MyBG0Wj2fs0pHmYKBGx:-MyHch-14hXHCVFrtXiB:b-wswmj1) ([raw file](https://github.com/harness/harness-core/blob/db40ed36ebdaa0f5a07205875e04f2ee419ff012/product/ci/scm/file/file.go#L452)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

Shouldn't NewObjectID be 2nd in the return param list? First one is commit id right?
</blockquote></details>

newobjectid is actually the commit id, it is very weird

**Comment by tphoney** [comment]
@mohitg0795 @shubham149 changes made where appropriate

**Review by mohitg0795 — APPROVED** [review]
:lgtm:

> Reviewed 6 of 14 files at r1, 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31165#-:-MyHiIwn6ma0Y_Ih-7Pq:bnfp4nl)* status: 14 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by shubham149 — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31165)* status: 14 file(s) to review.

**Review by shubham149 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31165)* status: 14 file(s) to review.

**Review by shubham149 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31165#-:-MyHpppR12QZqBwVxufB:bnfp4nl)* status: 14 file(s) to review.

**Review by shubham149 — APPROVED** [review]
> Reviewed 6 of 14 files at r1, 6 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31165)* status: 2 file(s) to review.

**Review by tphoney — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31165#-:-MyI5YG95iLsRL_tWDSq:bnfp4nl)* status: 2 file(s) to review.

**Review by tphoney — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31165#-:-MyI5rDwDKMS1ZHCpQlK:bnfp4nl)* status: 2 file(s) to review.

**Review by mohitg0795 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31165#-:-MyINkh9ERvCsb0Eyu4S:bnfp4nl)* status: 2 file(s) to review.

**Review by arpit3107 — APPROVED** [review]
> Reviewed 6 of 14 files at r1, 6 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31165)* status: 2 file(s) to review.

**Review by mohitg0795 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/31165#-:-MyIPHRiBzHCt_v0cAWs:bnfp4nl)* status: 2 file(s) to review.

**Pull request merged**